src/radix_tree.rs

```diff
@@ -0,0 +1,164 @@
+            if let Some(child) = maybe_child {
+                node = child;
+            } else {
+                return None.as_ref();
+            }
+        }
+
+        node.value.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::vwmap::{NamespaceFormat, NamespaceType};
+    use super::*;
+
+    #[test]
+    fn test_insert_and_get() {
+        let mut tree = RadixTree::new();
+
+        let namespace_descriptor_with_hash_1 = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 0,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
+            },
+            hash_seed: 1
+        };
+
+        let namespace_descriptor_with_hash_2 = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 10,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
+            },
+            hash_seed: 2
+        };
+
+        let namespace_descriptor_with_hash_3 = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 20,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
+            },
+            hash_seed: 3
+        };
+
+        tree.insert(b"A", namespace_descriptor_with_hash_1);
+        tree.insert(b"AB", namespace_descriptor_with_hash_2);
+        tree.insert(b"ABC", namespace_descriptor_with_hash_3);
+
+        assert_eq!(tree.get(b"A"), Some(&namespace_descriptor_with_hash_1));
+        assert_eq!(tree.get(b"AB"), Some(&namespace_descriptor_with_hash_2));
+        assert_eq!(tree.get(b"ABC"), Some(&namespace_descriptor_with_hash_3));
+        assert_eq!(tree.get(b"ABCD"), None);
+    }
+
+    #[test]
+    fn test_insert_and_get_empty_key() {
+        let mut tree = RadixTree::new();
+
+        let namespace_descriptor_with_hash = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 0,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
+            },
+            hash_seed: 1
+        };
+
+        tree.insert(b"", namespace_descriptor_with_hash);
+
+        assert_eq!(tree.get(b""), Some(&namespace_descriptor_with_hash));
+        assert_eq!(tree.get(b"A"), None);
+    }
+
+    #[test]
+    fn test_insert_and_get_long_key() {
+        let mut tree = RadixTree::new();
+
+        let namespace_descriptor_with_hash = NamespaceDescriptorWithHash {
+            descriptor: NamespaceDescriptor {
+                namespace_index: 0,
+                namespace_type: NamespaceType::Primitive,
+                namespace_format: NamespaceFormat::Categorical
+            },
+            hash_seed: 1
```

Comment:
Perhaps another nit-picky question here -- why are you hashing full structs in the first place? Why not just the numbers/hashes that would otherwise require re-computation? Surely overhead does not come from descriptor part?

Reply:
Beforehand we were doing 2 lookups, one from the descriptor & another for the hash seed. This way you do only one lookup where both are part of the same data structure. & in a radix tree which for a specific length of input is faster then the hashing.

Reply:
Great, thanks - yeah I guess this could also be a custom Hasher that would again result in just one lookup, but that's kind of what's happening anyways